eeps/eep-0079.md

```diff
@@ -0,0 +1,763 @@
+
+#### Validation
+
+A native record creation is validated at runtime against the
+native record definition. It fails with a `badrecord` error in the
+following cases:
+
+* There is no corresponding native-record definition.
+
+* The native-record definition is not visible at the call site (it is
+  not exported).
+
+* The native record create expression references the field FN which is
+  not defined (in the structure definition).
+
+* No value is provided for a field FN and the native-record definition
+  has no default value for FN.
+
+#### Native-record values and native-record definitions at runtime
+
+Now that we have seen how to define and create native records, it
+makes sense to specify how they behave at runtime, particularly in the
+context of Erlang's dynamic nature. Native records in this proposal
+are designed to support Erlang dynamism in a flexible way. The main
+scenarios:
+
+* Code upgrade:
+    * Native-record definitions can be upgraded (or even removed).
+
+* Distributed Erlang:
+    * Native records can travel between the Erlang nodes having different
+      versions of code.
+
+From now on let’s employ more precise terminology (when needed):
+*native-record definitions* and *native-record values*.
+
+When a native-record value is created, it “captures” key information
+from the current native-record definition, namely:
+
+* They fully qualified name of the native-record (module name and
+  native-record name)
+
+* Field names
+
+* Whether it is exported (through `-export_record`)
+
+The runtime needs the current native-record definition to perform
+create and update operations.
+
+However, to perform read operations on native-record values — accessing
+native-record fields and pattern matching over native-record values —
+the runtime does not consult the current native-record definition.
+
+To minimize the ambiguity of the next sections we would use more verbose wording:
+
+* The fields of a native-record value
+* The fields of a native-record definition
+* The native-record value is exported
+* The native-record definition is exported
+
+In the simplest case — a single Erlang node without code reloading,
+native-record values and native-record definitions would always be in
+sync.
+
+### Accessing native-record fields
+
+The syntax for accessing native-record fields is as follows:
+
+```erlang
+Expr#Name.Field
+Expr#Module:Name.Field
+```
+
+These expressions return the value of the specified field of the
+native-record value.
+
+An access operation fails with a `badfield` error if:
+
+* The native-record value is not visible at the call site (the
+  native record is not exported and is used outside of the defining
+  module).
+
+* The field `Field` is not defined in the native-record value.
+
+The expression can be used in guards — a guard would fail if the
+corresponding expression raises.
+
+The following syntax allows accessing field `Field` in any record:
+
+```erlang
+Expr#_.Field
```

Comment:
This creates a new use case because we could have two records `v1` and `v2` that share some fields, and use this syntax to access the common fields. It probably needs to be highlighted.